src/extension/authkey/src/main/java/org/geoserver/security/GeoServerAuthenticationKeyProvider.java

```diff
@@ -7,24 +7,73 @@
 package org.geoserver.security;
 
 import java.io.IOException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 import org.geoserver.config.util.XStreamPersister;
+import org.geoserver.platform.GeoServerExtensions;
 import org.geoserver.security.config.SecurityNamedServiceConfig;
 import org.geoserver.security.filter.AbstractFilterProvider;
 import org.geoserver.security.filter.GeoServerSecurityFilter;
 import org.geoserver.security.validation.SecurityConfigValidator;
+import org.geotools.util.logging.Logging;
+import org.springframework.beans.factory.DisposableBean;
+import org.springframework.scheduling.concurrent.CustomizableThreadFactory;
 
 /**
  * Security provider for auth-key authentication
  *
  * @author mcr
  */
-public class GeoServerAuthenticationKeyProvider extends AbstractFilterProvider {
+public class GeoServerAuthenticationKeyProvider extends AbstractFilterProvider
+        implements DisposableBean {
+
+    static Logger LOGGER = Logging.getLogger("org.geoserver.security");
+
+    // Use a counter to ensure a unique prefix for each pool.
+    private static final AtomicInteger poolCounter = new AtomicInteger();
+
+    private final GeoServerSecurityManager securityManager;
+
+    private final ScheduledExecutorService scheduler;
+    private final int autoSyncDelaySeconds;
+
+    /** @return a thread factory for the eviction thread */
+    private ThreadFactory getThreadFactory() {
+        CustomizableThreadFactory tFactory =
+                new CustomizableThreadFactory(
+                        String.format(
+                                "GeoServerAuthenticationKey-%d-", poolCounter.getAndIncrement()));
+        tFactory.setDaemon(true);
+        return tFactory;
+    }
+
+    /**
+     * @param securityManager the security manager
+     * @param autoSyncDelaySeconds the delay in seconds between each synchronization
+     */
+    public GeoServerAuthenticationKeyProvider(
+            GeoServerSecurityManager securityManager, int autoSyncDelaySeconds) {
+        this.securityManager = securityManager;
+        this.autoSyncDelaySeconds = autoSyncDelaySeconds;
+        this.scheduler = Executors.newScheduledThreadPool(1, getThreadFactory());
+
+        // schedule auto-sync thread
+        scheduler.scheduleAtFixedRate(
+                authKayMapperSyncTask,
+                autoSyncDelaySeconds,
+                autoSyncDelaySeconds,
+                TimeUnit.SECONDS);
```

Comment:
Shouldn't the delay be user controllable?

Reply:
The mappers are prototypes and the schedulers will be removed from memory within the filter configuration. Moreover the delay seconds can be eventually tweaked from the applicationContext and this is more a DevOps tuning configuration other than a user one IMHO.